### DISCUSSION TITLE
[Q&A] Turn off global model broadcasting in Swarm Learning

### DISCUSSION THREAD

**Discussion by jingyunc** [NVIDIA/NVFlare#2222] [Q&A] [answered]
### Python version (`python3 -V`)

3.8

### NVFlare version (`python3 -m pip list | grep "nvflare"`)

2.4.0rc3

### Operating system

Ubuntu 20.04

### Have you successfully run any of the following examples?

- [ ] hello-numpy-sag with simulator
- [ ] hello-pt with simulator
- [ ] hello-numpy-sag with POC
- [ ] hello-pt with POC

### Please describe your question

In the Swarm Learning example, is there an easy way to disable the global model broadcasting, so each site trains independently (except for the aggregator site) at each round? I would like to use the outputs as a baseline.

I noticed the associated learner (cifar10_model_learner) has function to save local model. Can the Swarm Learning workflow continue to run without the global model broadcasting at the end of each round?  Thanks.

**Comment by YuanTingHsieh** [comment]
@jingyunc you can comment out `broadcast_final_result` in this file: https://github.com/NVIDIA/NVFlare/blob/main/nvflare/app_common/ccwf/swarm_client_ctl.py

**Reply by jingyunc** [reply]
I commented all 3 places of broadcast_final_result in swarm_client_ctl.py. But there were two issue:

1. The first two places (Line 480 and 494) only got executed once, at the end of entire swarm learning. After I commented those two places out, the simulation job got stuck at that step can could not finish.
2. My goal was to disable the broadcasting after each aggregation round, is that triggered at the third place (Line 664)? During my test simulation, that place didn't got executed at all (I left log_info to mark the skipping of each broadcasting). Is there maybe a different function for broadcasting per aggregation round (rather than at the end)?

Thanks.

**Reply by yanchengnv** [reply]
broadcast_final_result isn't the place to change - it is called only after all rounds are finished to share the final results to all clients.

But I don't understand why you want to disable broadcasting (which is done by the _scatter method called by _end_gather) after each aggregation round. The whole point of the Swarming Learning is to let each client train on the last round's aggregated weights. So if this is disabled, then how do you want each client to train in the new round, and how is the aggregated result to be used?

**Reply by jingyunc** [reply]
Thanks @yanchengnv for the clarification. The reason we wanted to disable broadcasting is two-fold. First, we were hoping to implement our in-house FL method on NVFlare, which has a different model updating mechanism. Second, to compare with Swarm Learning and our in-house method, we need a baseline of "individual training", where each site trains on its own data without receiving global model.

Is it possible to turn off the per-round broadcasting in _scatter without interrupting the remaining workflow? Thanks.

**Reply by chesterxgchen** [reply]
For the baseline model "Individual training". you don't need to disable the broadcast to do so. 
You can simply initialize the model locally at each site. Make sure all size initialized the same way.  Then make total number round = 1
Then in the client code, ignore the global model received broadcast to you. and use the locally initialized model to start. 
That will give you what you want easily

**Reply by jingyunc** [reply]
Can you please elaborate how to initialize the model locally at each site? Can this be done in real-world mode only or it can be done in simulation mode as well?

**Reply by chesterxgchen** [reply]
For example, suppose the training code is CIFAR10 train.py 

https://github.com/NVIDIA/NVFlare/blob/main/examples/hello-world/step-by-step/cifar10/code/fl/train.py

L95 is flare.receive() which means receive the global model from controller (from another client). 

https://github.com/NVIDIA/NVFlare/blob/main/examples/hello-world/step-by-step/cifar10/code/fl/train.py#L95

You can simply replace that with 

input_model = local_init_model() 

where local_init_model() could be something like

def local_init_model(): 
    net = Net()
    input_model =  FLModel( params = net.state_dict(), params_type = "FULL") 
    return input_model

by doing so, you essentially ignore the model received from controller, and always use the local model. 
By set the total number of round = 1, you can save trained the final model locally.  so every site/client has the centralized model with their own dataset. 

I hope that's what you are looking for.

**Reply by YuanTingHsieh** [reply, marked as answer]
@jingyunc For getting the baseline metrics of each individual training, you can just run train script on each client individually, don't need to use NVFlare (this is just as regular ML/DL application)

**Comment by chesterxgchen** [comment]
@yanchengnv  can you help?
